pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,169 @@
+package supportconfigs
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	v1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	authzv1 "k8s.io/api/authorization/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const (
+	// Endpoint The endpoint that this URL is accessible at - used for routing and for SARs
+	Endpoint             = "/v1/generateSUSERancherSupportConfig"
+	cspAdapterNamespace  = "cattle-csp-system"
+	cspAdapterConfigmap  = "csp-config"
+	tarContentType       = "application/x-tar"
+	unauthenticatedError = "unauthenticaed"
+	forbiddenError       = "forbidden"
+)
+
+type GeneratorHandler struct {
+	ConfigMaps           v1.ConfigMapInterface
+	SubjectAccessReviews authv1.SubjectAccessReviewInterface
+	ConfigHandler        supportConfigHandler
+}
+
+func NewGeneratorHandler(scaledContext *config.ScaledContext) GeneratorHandler {
+	return GeneratorHandler{
+		ConfigMaps:           scaledContext.Core.ConfigMaps(cspAdapterNamespace),
+		SubjectAccessReviews: scaledContext.K8sClient.AuthorizationV1().SubjectAccessReviews(),
+		ConfigHandler:        supportConfigHandler{},
+	}
+}
+
+func (h *GeneratorHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	err := h.authorize(request)
+	if err != nil {
+		switch err.Error() {
+		case unauthenticatedError:
+			writer.WriteHeader(http.StatusUnauthorized)
+			_, _ = writer.Write([]byte(err.Error()))
+		case forbiddenError:
+			writer.WriteHeader(http.StatusForbidden)
+			_, _ = writer.Write([]byte(err.Error()))
+		default:
+			writer.WriteHeader(http.StatusInternalServerError)
+			logrus.Errorf("Failed to authorize user with error: %s", err.Error())
+		}
+		return
+	}
+	logrus.Infof("Generating supportconfig")
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("Done Generating supportconfig")
+	if err != nil {
+		logrus.Errorf("Error when generating support config %v", err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "filename=\"supportConfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Infof("wrote %v bytes in archive response", n)
+}
+
+// authorize checks to see if the user can get the non-resource url of Endpoint. If not, returns an error, otherwise nil
+func (h *GeneratorHandler) authorize(r *http.Request) error {
```

Comment:
Suggestion - it might help with consistency with k8s/rancher to use some of the other `Authorize` methods for inspiration -

https://pkg.go.dev/k8s.io/apiserver/pkg/authorization/authorizer#AuthorizerFunc.Authorize
https://github.com/rancher/rancher/blob/8a31d3cc1bf0830439d4926cd06d8fccfb9c2c54/pkg/api/steve/aggregation/auth.go#L50
https://github.com/rancher/rancher/blob/8a31d3cc1bf0830439d4926cd06d8fccfb9c2c54/pkg/tunnelserver/mcmauthorizer/tunnel.go#L108

where they return multiple values, one to indicate an error and another to indicate an auth decision. This would be cleaner to parse above when deciding what HTTP error to return. Another thing they do is treat missing the token/auth context as the same as a forbidden error instead of a separate unauthenticated error, which also simplifies the code.

Reply:
Good point. I have changed this to return a bool and an error. In addition, I no longer parse for unauthenticated cases. Now if authorized returns false or an error then we still give back a `403`. We do log an error rancher side for the error cases, but that isn't exposed to the user.